packages/components/src/text-input/hooks/use-text-input-state.js

```diff
@@ -0,0 +1,494 @@
+/**
+ * External dependencies
+ */
+import { isNil, noop } from 'lodash';
+
+/**
+ * WordPress dependencies
+ */
+import {
+	useCallback,
+	useEffect,
+	useMemo,
+	useRef,
+	useState,
+} from '@wordpress/element';
+
+/**
+ * Internal dependencies
+ */
+import { useBaseDragHandlers } from './use-text-input-state-utils';
+import { isValueNumeric } from '../../utils/values';
+import { add, roundClampString, subtract } from '../../utils/math';
+import { mergeEventHandlers } from '../../utils/events';
+import { normalizeArrowKey } from '../../utils/keyboard';
+import { useControlledValue, usePropRef } from '../../utils';
+
+/**
+ * @template T
+ * @param {Object} props
+ * @param {T} [props.value]
+ * @return {Array<any|Function>} The controlled value and the value setter.
```

Comment:
We should fix this return type. I think it should be something like `[T, (value: T) => void, () => void]`
```suggestion
 * @return {[T, (value: T) => void, () => void]} The controlled value and the value setter.
```